### PR TITLE
Split request/response model types; tolerant responses via all-Option fields

### LINKED ISSUE #314: Split request/response model types; tolerant responses via all-Option fields (supersedes #312)
## Problem

#312 adopted tolerant response deserialization by sweeping `#[serde(default)]` across every model field (implemented in #313). Review of that PR surfaced a structural flaw in the mechanism: `#[serde(default)]` on a required `T` field fabricates a value (`""`/`0`/`false`) that is indistinguishable from a genuine server-sent value.

- **Write-back hazard.** `clickhouse-cloud-api` is published to crates.io, and some types flow in both directions — `PostgresInstanceConfig` is literally the same type in the `postgres_instance_config_get` response and the `postgres_instance_config_post`/`patch` body. A consumer doing get → tweak → post silently persists fabricated values for any field the server dropped. #313 documents this as the "GET→write-back caveat" rather than closing it.
- **Semantic loss.** "Server sent `0`" and "server dropped the field" become the same value — in `--json` output, in `print_human`, and in any consumer logic.

The goal of #312 stands unchanged: a server-dropped response field must not fail the response — multiple teams evolve the Cloud API independently, and each strict response field is a latent outage point. Only the mechanism changes.

## Decision

Split request and response model types, and express response tolerance in the type system instead of via serde attributes:

- **Request types: unchanged.** Required fields stay `T` (the type system enforces "strict in what we send"), optional/nullable fields stay `Option<T>` + `skip_serializing_if`. No `#[serde(default)]` anywhere.
- **Response types: every field is `Option<T>`.** A missing key deserializes to `None` natively (no serde attribute needed), and — unlike `serde(default)`, which only fills a *missing* key — an explicit JSON `null` also lands as `None`, closing a residual failure mode #313 explicitly left open (e.g. `null` on a `Vec` field). Nothing is fabricated; absence is explicit and each caller decides what absence means at the point of use.
- **Enums and unions: unchanged.** String enums keep their `Unknown(String)` catch-alls; object unions stay on `discriminated_union!` with lossless `Unknown(Value)`.

## Current shape of the code (measured, as of the #311 branch)

- The operation layer is already ~split: 50 request-body types vs 49 response types, and only **2 top-level types** cross directions (`ClickStackSource`, `PostgresInstanceConfig`).
- **75 nested structs** are transitively shared between the request tree and the response tree. These are the types that need splitting into a request variant (strict) and a response variant (all-`Option`). Concentrated in ClickStack (chart configs, sources, dashboard containers, on-click targets), ClickPipes (pipe settings, table mappings, destination columns, scaling), and Postgres (`PostgresInstanceConfig`, `PgConfig`, `PgBouncerConfig`), plus a few shared leaves (`IpAccessListEntry`, `ResourceTagsV1`, `RBACPolicyTags`, `CustomPrivateDnsMapping`).
- The full response tree is ~330 types — the all-`Option` conversion touches all of them, not just the 75 shared ones.
- The spec itself already splits in places, so there is precedent to follow: ClickPipes `ClickPipePost*`/`ClickPipePatch*`/`ClickPipeMutate*` vs the response shapes, and ClickStack `*Input` vs `*Output`/`*Response`.

<details>
<summary>The 75 shared structs</summary>

ClickPipeBigQueryPipeSettings, ClickPipeBigQueryPipeTableMapping, ClickPipeDestinationColumn, ClickPipeDestinationTableDefinition, ClickPipeDestinationTableEngine, ClickPipeFieldMapping, ClickPipeKafkaOffset, ClickPipeMongoDBPipeSettings, ClickPipeMongoDBPipeTableMapping, ClickPipeMySQLPipeSettings, ClickPipeMySQLPipeTableMapping, ClickPipePostgresPipeSettings, ClickPipePostgresPipeTableMapping, ClickPipeScaling, ClickPipeSettings, ClickStackAggregatedColumn, ClickStackAlertChannelEmail, ClickStackAlertChannelWebhook, ClickStackBackgroundChart, ClickStackBarBuilderChartConfig, ClickStackBarRawSqlChartConfig, ClickStackBetweenColorCondition, ClickStackCASLPermission, ClickStackCategoricalBarBuilderChartConfig, ClickStackCategoricalBarRawSqlChartConfig, ClickStackDashboardContainer, ClickStackDashboardContainerTab, ClickStackEqualityColorCondition, ClickStackEventPatternsChartConfig, ClickStackFilter, ClickStackFilterSettingsColumn, ClickStackHeatmapChartConfig, ClickStackHeatmapSelectItem, ClickStackHighlightedAttributeExpression, ClickStackLineBuilderChartConfig, ClickStackLineRawSqlChartConfig, ClickStackLogSource, ClickStackLogSourceMetadataMaterializedViews, ClickStackMarkdownChartConfig, ClickStackMaterializedView, ClickStackMetricSource, ClickStackMetricSourceFrom, ClickStackMetricTables, ClickStackNumberBuilderChartConfig, ClickStackNumberFormat, ClickStackNumberRawSqlChartConfig, ClickStackNumericColorCondition, ClickStackOnClickDashboard, ClickStackOnClickExternal, ClickStackOnClickFilterTemplate, ClickStackOnClickSearch, ClickStackOnClickTargetIdVariant, ClickStackOnClickTargetTemplateVariant, ClickStackPieBuilderChartConfig, ClickStackPieRawSqlChartConfig, ClickStackPromqlSource, ClickStackQuerySetting, ClickStackSavedFilterValue, ClickStackSavedSearchFilter, ClickStackSearchChartConfig, ClickStackSelectItem, ClickStackSessionSource, ClickStackSourceFilterSettings, ClickStackSourceFrom, ClickStackTableBuilderChartConfig, ClickStackTableRawSqlChartConfig, ClickStackTraceSource, ClickStackTraceSourceMetadataMaterializedViews, CustomPrivateDnsMapping, IpAccessListEntry, PgBouncerConfig, PgConfig, PostgresInstanceConfig, RBACPolicyTags, ResourceTagsV1

(Shared *enums* — ~94 of them — do not need splitting: value enums with `Unknown` catch-alls are direction-safe as-is.)
</details>

## What needs to be done

**Models (`models.rs`)**
- Split the 2 top-level bidirectional types and the 75 shared nested structs into request/response variants. Pick one naming convention up front and apply it uniformly (the spec's own precedents are `*Input`/`*Output` and `Mutate*`; whatever is chosen must be encodable in the analyzer's mapping, see below).
- Convert every field of every response-tree type (~330 types) to `Option<T>`. Remove all `#[serde(default)]` (it becomes meaningless on `Option` fields and is banned on request fields by the write-back argument).
- Response types must keep `Serialize` — the CLI's `--json` output and `print_human` both serialize response models. Decide deliberately whether absent fields serialize as `null` or are omitted (`skip_serializing_if`); omission is probably right so `--json` output distinguishes "absent" without inventing `null` noise, but it should be an explicit decision with a test pinning it.
- Consider `From<FooResponse> for FooRequest`-style conversion helpers (or `TryFrom` where required request fields may be absent in the response) for consumers who legitimately do get → modify → write. The point of the split is that this conversion is *explicit* — absent fields must be consciously resolved — not that it's impossible.

**Analyzer (`clickhouse-openapi-analyzer`)**
- Requiredness resolution becomes direction-aware: for request-position schemas, `required[]`/PATCH-all-optional/description-heuristic rules apply as today; for response-position schemas, every field is expected to be `Option<T>` and optionality findings are suppressed (they'd otherwise fire on every response field). Field *presence* checks (missing/extra fields) remain the drift signal in both directions and must keep working.
- A spec schema used in both positions now maps to **two** Rust types. The schema↔type mapping in `compare.rs`/`rust_inventory.rs` must resolve both, each checked under its direction's rule. The naming convention chosen for the split is what makes this mechanical — encode it, don't hand-map.
- Exemption keys (`optionality_exemptions`, `extra_field_exemptions`, `deprecated_field_exemptions`) are keyed by `RustStructName` — audit existing entries for structs being split and re-key to the correct variant(s). Same for `partial_required_schemas` semantics (now request-side only).
- Drop #313's `serde(default)` inventory tracking and `model_fields_missing_serde_default()` if that commit was carried; it enforces the superseded policy.

**Policy enforcement**
- Replace #313's `every_model_field_carries_serde_default` test with the equivalent structural enforcement for the new policy: every field of every response-reachable type is `Option<T>`, empty exception list. This likely needs the analyzer to expose request/response tree membership (it already inventories client method signatures, which define the roots).

**Deprecated-field hiding and meta**
- `meta.rs::DEPRECATED_FIELDS` and `#[cfg(feature = "deprecated-fields")]` markers are keyed per struct — split structs double the bookkeeping. Regenerate with `scripts/regenerate-deprecated-fields.py` and verify both feature configurations compile (`cargo check --workspace --all-features`).

**CLI (`crates/clickhousectl`)**
- Every read of a response field must handle `None`: `print_human` paths, `tabled` list views, and any logic branching on response values. Display code already handles `Option` widely (~69 sites in `cloud/commands.rs`), so this is mechanical — render absence as `-`/empty, never `unwrap()`/`expect()` on a response field.
- `postgres.rs` constructs `PostgresInstanceConfig` from user input for `config set` — it must move to the request variant.

**Tests and docs**
- `models_test.rs` round-trips, `client_test.rs`, integration tests, and CLI parsing/builder tests all touch response field access — expect broad mechanical churn. Add targeted tests: missing-key → `None`, explicit-`null` → `None`, and the serialization (`null` vs omit) decision.
- Rewrite the tolerant-deserialization policy in CLAUDE.md/AGENTS.md ("Optionality and deserialization tolerance" section) and the crate-docs section #313 added. The current text argues for the `serde(default)` policy and must not survive as the policy record, for the same reason this issue supersedes #312.

## What carries over from #313 (being closed)

Cherry-pick, don't rewrite — these are needed under this approach too, and all-`Option` makes them *more* necessary: an all-`Option` variant matches any JSON object, so untagged shape-matching degenerates completely and explicit discriminator dispatch is the only thing that works.

- `125ad7a` — `ClickStackAlertChannel` → `discriminated_union!` dispatch on `type`.
- `d57839e` — six chart-config Builder/RawSql unions dispatched on `configType` presence, including the `none unless` disqualifying-key guard.
- The union-semantics parts of `6d1eb1b` — malformed-payload-for-known-discriminator → lossless `Unknown(Value)` fallback, and the `Default`-round-trips-to-same-variant invariant test. This commit mixes those fixes with sweep-dependent changes, so it needs splitting rather than a clean cherry-pick.

The `discriminated_union!` macro itself lives in #311 and merges independently.

## Keep in mind

- **What this still does not cover:** a plain struct field whose *type* changes server-side (key present, wrong shape) still fails deserialization — `Option<T>` doesn't absorb that any more than `serde(default)` did. Unions absorb it via the `Unknown` fallback; enums via their catch-alls. Document this residual honestly, as #313 did.
- **Discriminator strictness survives the sweep by construction:** `discriminated_union!` dispatches on the raw JSON value, not on struct fields, so all-`Option` variant structs don't weaken dispatch. The #313 adversarial-review findings (RawSql-without-`configType` misroute, `Default` round-trip mismatch) remain the reference for the failure modes to test.
- **Response-side optionality drift becomes invisible** to the analyzer by design (everything is `Option`). That signal was low-value; presence/absence drift is what matters and is retained. Say so in the analyzer docs rather than leaving it as an apparent gap.
- **`VALUES` consts, beta lists, and enum coverage checking** are unaffected.
- Sequencing and commit slicing are deliberately out of scope here — to be planned by whoever picks this up.

Supersedes #312. Replaces the approach in #313.

### PR DESCRIPTION
**Pull request by sdairs**
Closes #314

**Stacked on #311** — base is `issue-308-clickstack-drift`; retarget to `main` after #311 merges. Replaces #313 (closed), whose `serde(default)` sweep fabricated values indistinguishable from server-sent ones.

## What and why

A server-dropped or `null` response field must never fail deserialization — each strict response field is a latent outage point. Tolerance now lives in the type system instead of serde attributes:

- **Request types: strict.** Required fields are `T`, optional are `Option<T>` + `skip_serializing_if`. The compiler enforces "strict in what we send".
- **Response types: all-`Option`.** Every field of every type reachable from a `Client` return type (339 types) is `Option<T>`. Missing key *and* explicit `null` both land as `None` — nothing is fabricated, so "server sent `0`" and "server dropped the field" stay distinguishable.
- **Enums/unions unchanged**: `Unknown` catch-alls absorb unrecognized values and (new in this PR) malformed payloads under a recognized discriminator.

## How

- **Naming:** a schema used in both directions splits into `{Name}` (request) + `{Name}Response` — 88 pairs, more than the 75+2 the issue measured (agents verified reachability from `client.rs` rather than trusting the list). Unions whose variant structs split are duplicated (`discriminated_union!` invocation per direction).
- **Serialization pinned:** absent response fields are **omitted** from `--json`/`print_human` output, never emitted as `null` (`skip_serializing_if` on every response field, test-enforced).
- **Write-back is explicit:** new `convert.rs` with `TryFrom<{Name}Response>`/`From` conversions (top-level: `PostgresInstanceConfig`, `ClickStackSource`); a fallible conversion returns `MissingRequiredFields` naming the missing wire fields. This closes #313's GET→write-back caveat instead of documenting it.
- **`serde(default)` is banned** in `models.rs`, enforced by test via the analyzer.
- **Analyzer is direction-aware:** requiredness rules apply in request position only; response-side optionality findings are suppressed by design (presence and enum-value drift remain checked in both directions); response-position schemas resolve to `{Name}Response` when it exists. New `response_tree()` API backs the enforcement tests.
- **Union hardening (carried from #313 + extended):** `none unless "connectionId" | "sqlTemplate"` guards on the chart-config unions (mandatory now — all-`Option` variants are total, so a dropped discriminator would otherwise silently retype a RawSql payload); known-discriminator malformed payloads degrade to lossless `Unknown(Value)`; every union `Default` round-trips to the same variant, with the covered list structurally required to equal the manual `impl Default` blocks.
- **CLI:** no `unwrap`/`expect` on response fields; absence renders as `-` (`or_absent`); `postgres config set` builds the request variant; delete `--json` no longer fabricates `status`/`requestId`.

## Residual (documented, intentional)

A key that is *present* with a changed type still fails a plain struct field — `Option<T>` absorbs absence and `null`, not shape changes. Enums/unions absorb those via catch-alls. Spec conformance stays the drift job's responsibility.

## Breaking changes (library consumers, pre-1.0)

- All response-tree fields are now `Option<T>`; response-position types returned by `Client` methods are the `{Name}Response` variants where split.
- `discriminated_union!` grammar: bare `none =>` arms no longer parse; absence-discriminated unions must name disqualifying keys.
- CLI `DeleteResponse` fields are `Option`.

## Verification

Every commit is independently green (per-commit verifier agents re-ran all gates): `cargo test --workspace`, clippy `-D warnings` workspace-wide, `cargo fmt --all --check`, `cargo check --workspace --all-features`, `python3 -m unittest discover -s scripts/tests`, and `python3 scripts/check-openapi-drift.py --dry-run` → 0 actionable findings against the live spec. A fresh-context adversarial review of the full stack found no high/medium defects; its three low-severity findings are fixed in the final commit. Enforcement tests were mutation-tested (strictening a response field / removing a covered union fails the corresponding test with the exact offender named).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by sdairs** [comment]
bugbot review

File: crates/clickhousectl/src/cloud/postgres.rs [hunk 1]
```diff
@@ -631,7 +679,16 @@ pub async fn postgres_update(
...
-        Some(merge_tags(&current.tags, &add, opts.remove_tag))
+        // An omitted `tags` in the response means "no tags to merge with"; a
+        // returned tag without a key cannot be sent back, so say so rather
+        // than dropping it from the merged set.
+        let existing = current
+            .tags
+            .unwrap_or_default()
+            .into_iter()
+            .map(ResourceTagsV1::try_from)
+            .collect::<Result<Vec<_>, _>>()?;
```
[the remark this answers is not in the document]
Comment (sdairs) [reply crates/clickhousectl/src/cloud/postgres.rs]: Fixed in 206a45e. `postgres_update` now refuses the merge when the response omits `tags` (new `merge_response_tags` helper errors on `None`), so a partial response can no longer wipe existing tags via the full-replacement PATCH. A genuine empty list (`"tags": []`) still merges normally. Unit tests cover both paths plus the keyless-tag refusal.

File: crates/clickhousectl/src/cloud/commands.rs [hunk 2]
```diff
@@ -900,7 +928,7 @@ pub async fn service_delete(
             loop {
                 tokio::time::sleep(std::time::Duration::from_secs(5)).await;
                 let svc = client.get_service(&org_id, service_id).await?;
-                let state = svc.state.to_string();
+                let state = or_absent(svc.state.as_ref());
```
[the remark this answers is not in the document]
Comment (sdairs) [reply crates/clickhousectl/src/cloud/commands.rs]: Fixed in 7b8ec45. An absent `state` during the stop-polling loop is now an error ("the API response omitted the service state while waiting for the service to stop") instead of polling forever. The pre-stop check keeps its existing behavior — absent state skips the stop attempt and lets the delete call decide. Covered by unit tests on the extracted `classify_stop_poll_state` helper.

File: crates/clickhousectl/src/cloud/postgres.rs [hunk 3]
```diff
@@ -264,6 +265,27 @@ fn load_json_file<T: DeserializeOwned>(path: &Path) -> Result<T, Box<dyn std::er
...
+/// to an empty object here — explicitly, at the point of use.
+fn instance_config_from_json(
+    doc: &serde_json::Value,
+) -> Result<PostgresInstanceConfig, Box<dyn std::error::Error>> {
+    let section = |key: &str| {
+        doc.get(key)
+            .cloned()
+            .unwrap_or_else(|| serde_json::json!({}))
+    };
+    Ok(PostgresInstanceConfig {
```
[the remark this answers is not in the document]
Comment (sdairs) [reply crates/clickhousectl/src/cloud/postgres.rs:288]: Fixed in bd7a37e. `instance_config_from_json` now rejects any non-object root ("configuration document must be a JSON object") before section extraction, so a scalar/array/`null` document errors instead of sending `{"pgConfig": {}, "pgBouncerConfig": {}}`. Object roots keep the documented behavior: an omitted section key still fills `{}`. Unit tests cover all non-object root types.

File: crates/clickhousectl/src/cloud/commands.rs [hunk 4]
```diff
@@ -2808,14 +2885,21 @@ pub async fn key_create(
...
-        if !resp.key_secret.is_empty() {
-            println!("  Key Secret: {}", resp.key_secret);
+        if !key_secret.is_empty() {
+            println!("  Key Secret: {}", key_secret);
         }
-        if !resp.key_id.is_empty() || !resp.key_secret.is_empty() {
+        if !key_id.is_empty() || !key_secret.is_empty() {
             println!();
             println!("Save the key secret now — it will not be shown again.");
         } else {
```
[the remark this answers is not in the document]
Comment (sdairs) [reply crates/clickhousectl/src/cloud/commands.rs]: Fixed in 7b8ec45. The pre-hashed branch is now gated on `request.hash_data.is_some()` as suggested. For a generated-key request, a response missing `keyId` or `keySecret` is now an error (before any success output) that names the possibly-created key so the user can list/delete it. Per the split policy, an explicitly-returned empty string counts as sent, not absent. Covered by unit tests on the extracted `resolve_key_create_material` helper.

File: crates/clickhousectl/src/cloud/service_query.rs [hunk 5]
```diff
@@ -57,13 +62,17 @@ pub async fn ensure_service_query_setup(
     };
 
     let key_response = client.create_api_key(org_id, &key_request).await?;
```
[the remark this answers is not in the document]
Comment (sdairs) [reply crates/clickhousectl/src/cloud/service_query.rs:83]: Fixed in 542cf57. All post-creation validation failures now do a best-effort `delete_api_key` before returning: absent `keyId`/`keySecret`, and an absent endpoint `id` after a successful bind. `key.id` is resolved first since it is the handle every cleanup needs — when it is absent there is nothing actionable, which is now documented at the check. One correction to the finding: at the `keyId`/`keySecret` checks the key is not yet bound (binding happens afterwards), but the orphaned-key-per-retry leak was real in all cases. Two new wiremock tests assert exactly one `DELETE /keys/{uuid}` is sent, nothing is persisted, and the CLI exits non-zero; both were confirmed to fail against the pre-fix code.

File: crates/clickhouse-openapi-analyzer/src/openapi.rs [hunk 6]
```diff
@@ -171,12 +197,105 @@ impl OpenApiInventory {
...
         }
         Ok(())
     }
+
+    /// Classifies every named schema as request-position and/or
+    /// response-position by walking `$ref`s from each operation's parameters
+    /// and request body (request roots) and responses (response roots), then
+    /// taking the transitive closure through the schema reference graph. A
+    /// schema can be in both positions.
+    fn collect_schema_positions(&mut self, spec: &Value) {
```
[the remark this answers is not in the document]
Comment (sdairs) [reply crates/clickhouse-openapi-analyzer/src/openapi.rs:215]: Fixed in 0719f04. Position classification now resolves local refs to `#/components/responses/`, `#/components/requestBodies/` and `#/components/parameters/` (any non-schema component, including nested component→component hops) and continues scanning through the resolved JSON, with a visited set breaking ref cycles. Note this is latent robustness: the current spec has zero such components, and the fix was confirmed behavior-preserving — the analyzer report over the vendored snapshot is byte-identical before and after. New unit tests cover response/requestBody/parameter component refs and cycle termination.

File: crates/clickhousectl/src/cloud/commands.rs [hunk 7]
```diff
@@ -856,18 +880,20 @@ pub async fn service_create(
     let org_id = resolve_org_id(client, opts.org_id.as_deref()).await?;
 
     let response = client.create_service(&org_id, &request).await?;
-    let svc_id = response.service.id.to_string();
+    let svc_id = or_absent(response.service.as_ref().and_then(|svc| svc.id));
```
[the remark this answers is not in the document]
Comment (sdairs) [reply crates/clickhousectl/src/cloud/commands.rs:926]: Fixed in 7b8ec45. The "Run SQL with ... --id ..." hint (and its provisioning follow-up line) is now only printed when the service id is present. Kept it a conditional print rather than an error: the create succeeded and the one-time password still has to be shown, so suppressing the unusable hint is the right failure mode here.

**Comment by sdairs** [comment]
Deep review of the full stack (19 commits, reviewed against `97ac80d` as the true PR boundary — the `client.rs` endpoint additions and union groundwork in the GitHub diff belong to #311). I ran the full gate suite locally rather than trusting the PR description: `cargo test` for both crates (480+ and 506 tests), clippy `-D warnings`, `cargo fmt --check`, `cargo check --workspace --all-features`, the Python unittests, and `check-openapi-drift.py --dry-run` → 0 actionable drift against the live spec. I also independently mutation-probed the union dispatch guard semantics with adversarial payloads.

**Verdict: approve.** The core design — tolerance in the type system rather than serde attributes — is strictly better than the #313 `serde(default)` sweep it replaces, because it preserves the absence/presence distinction instead of fabricating values. Enforcement is structural (derived from `client.rs` return types with a vacuity guard), not conventional.

What impressed me most:

1. **The `none unless` guard is correct and load-bearing.** With all-`Option` variant structs the absence arm is total, so a raw-SQL chart config with a server-dropped `configType` would silently retype to an empty builder config, discarding `connectionId`/`sqlTemplate`. I probed all five adversarial cases (dropped discriminator, empty object, malformed payload under a known discriminator, unknown discriminator, non-string discriminator) — every one routes exactly as documented, and the macro rustdoc honestly owns the deliberate "absent vs non-string" conflation.

2. **The union `Default` round-trip invariant is enforced structurally** — the test's covered list must equal the analyzer-inventoried set of manual `impl Default` blocks, so a new union can't silently escape it. The two fixes this produced (`BackupBucket`/`ClickStackWebhook` naming their own discriminator wire value; `ClickStackAlertChannelEmailType` defaulting to `Email`) are exactly the class of bug that otherwise ships silently.

3. **The review-round commits are real fixes, not cosmetics:** `merge_response_tags` refusing to merge on an omitted `tags` (prevents wholesale tag deletion via PATCH-replaces-set semantics); `classify_stop_poll_state` failing on absent state (the old code spun the timeout-less poll forever); `resolve_key_create_material` reading the mode from the request rather than inferring from the response (prevents printing success while the one-time secret is lost); `ensure_service_query_setup` resolving `key.id` first and routing every later failure through key deletion (kills the orphaned-key-per-retry leak, with wiremock coverage end-to-end).

Minor findings, none blocking:

- **`ClickStackAlertChannelEmailType`'s `#[default] Email` is only protected at one remove.** The union-level invariant test covers it, but the field enum is shared by both alert-channel variants (both declare `["webhook","email"]`), so a future spec change adding a variant could re-open the misroute. A one-line comment on the enum itself — "default must match the email variant's discriminator" — would make it self-defending.
- **`DeleteResponse` (CLI) keeps `rename_all = "camelCase"`** while the library now mandates explicit renames for greppability. Deliberate and harmless (the analyzer only parses `models.rs`); noting only for completeness.
- **Coverage gap worth restating for the record:** the all-`Option` migration touched every integration suite's assertions, but only Postgres CDC runs in CI. A real response-shape regression on a third-party pipe or ClickStack endpoint won't surface until someone runs the manual suite.

Things I specifically checked that are NOT problems: no fabrication anywhere on the response path (`DeleteResponse` no longer invents `0.0`; `key create`/`service create`/`service_query` resolve absence explicitly; `postgres_reset_password` falls back to the password we sent, not a placeholder); `--json` emits exactly the API's key set (pinned by test); `serde(default)` is genuinely gone from `models.rs` including the container-level dodge; the `response_position_type` guard against a spec schema literally named `{Name}Response`; convert.rs is uniform (collects all missing fields, wire names, nested at the right level — including the literal-space `"exponential histogram"` wire name, which matches the spec).

The main risk — 88 hand-maintained split pairs — is inherent to the design and mitigated as well as it can be by the direction-aware analyzer. Ship it (after retargeting to `main` once #311's base merges, per the description).

**Comment by sdairs** [comment]
Deep review of the current head (`0719f04`). The request/response split and analyzer design are strong, but I found five omission-handling paths that should be fixed before merge.

### Findings

1. **[P1] Preserve existing query-endpoint keys when `openApiKeys` is absent**  
   `crates/clickhousectl/src/cloud/service_query.rs:151-158`

   A successful endpoint GET with `openApiKeys` omitted or `null` is coerced to an empty list. The following upsert is a full replacement, so provisioning a new key can silently revoke every existing query-endpoint key. Treat `None` as an incomplete response and abort/clean up; only a genuine 404 should initialize an empty list.

2. **[P2] Do not delete a usable key over an unused endpoint ID**  
   `crates/clickhousectl/src/cloud/service_query.rs:117-125`

   The endpoint upsert has already succeeded, but an omitted response `id` causes the newly bound key to be deleted, leaving a dangling UUID in `openApiKeys`. `ServiceQueryKey.endpoint_id` is never consumed outside persistence/tests. Persist without depending on the echoed ID, refetch it, or explicitly roll back the binding before deleting the key.

3. **[P2] Distinguish generated-password responses**  
   `crates/clickhousectl/src/cloud/commands.rs:2221-2230`

   When neither hash flag is supplied, the empty request asks the API to generate a new password. If the response omits that one-time password, the command incorrectly reports “Password hash updated” and exits successfully, leaving the user without the generated credential. Resolve behavior from the request mode and report an incomplete response when generation was requested.

4. **[P2] Do not render missing one-time credentials as `-`**  
   `crates/clickhousectl/src/cloud/commands.rs:897-910`  
   `crates/clickhousectl/src/cloud/postgres.rs:676-678`

   A created ClickHouse or Postgres service whose response omits `password` is reported as successful with `Password: -`, even though that password is shown only once. Emit a prominent incomplete-response warning/error with recovery instructions instead of presenting the placeholder as a credential.

5. **[P2] Validate generated key material before branching on output mode**  
   `crates/clickhousectl/src/cloud/commands.rs:2961-2974`

   `resolve_key_create_material` runs only for human output. Explicit `--json` and automatically detected coding-agent invocations bypass it, print an incomplete generated-key response, and exit successfully despite losing the one-time secret. Move validation before the output-mode branch while preserving raw JSON when the response is complete.

### Verification

I ran:

- `cargo test --workspace`
- `cargo clippy --workspace --all-targets -- -D warnings`
- `cargo fmt --all --check`
- `cargo check --workspace --all-features`
- `python3 -m unittest discover -s scripts/tests -p 'test_*.py'`
- `python3 scripts/check-openapi-drift.py --dry-run` against the live spec: **0 actionable findings**

The worktree remained clean.

The current Copilot comment about `enum_wire_value` causing a compile failure is a false positive: Rust permits these partial moves because subsequent accesses use different fields, and the full workspace compiles successfully.

**Verdict: request changes.** The remaining issues can revoke existing credentials or lose one-time secrets under the exact partial-response conditions this PR is intended to make safe.

**Comment by sdairs** [comment]
All five findings were valid and are now fixed, one commit each, with tests confirmed to fail against the pre-fix code:

1. **[P1] `openApiKeys` merge** — fixed in 07ba6f6. `bind_query_endpoint` now refuses a 200 GET whose `result` or `openApiKeys` is absent (the error names the field and the revocation consequence); only a genuine 404 initializes an empty list, and an explicit `"openApiKeys": []` still merges. The existing cleanup arm discards the just-provisioned key on this path — pinned end-to-end by a wiremock test asserting zero endpoint POSTs, exactly one `DELETE /keys/{uuid}`, nothing persisted, non-zero exit.

2. **`endpoint_id` deletion** — fixed in 45941b7. `ServiceQueryKey.endpoint_id` is now `Option<String>` (CLI-internal storage, old credential files still deserialize) and an absent echoed `id` persists the credentials with `endpoint_id: None` instead of deleting the freshly bound key. Wiremock test: upsert 200 without `id` → exit 0, no `DELETE /keys/`, key pair persisted.

3. **Generated-password mode** — fixed in a927957 + corrected in 049aeca. `service_reset_password` now resolves the mode from the request via a `resolve_reset_password_outcome` helper that runs **before** the `--json`/human branch, so both modes fail when a requested generated password is omitted (explicit `""` counts as sent, mirroring `resolve_key_create_material`). The follow-up commit matters: per the spec, `newDoubleSha1Hash` alone is *ignored* and a password is still generated, so mode is keyed on `newPasswordHash` alone — the initial fix (and the finding as scoped) would have silently discarded the generated password for a `--new-double-sha1-hash`-only reset. Unit tests cover all mode×presence cells plus that flag combination; wiremock tests cover both output modes.

4. **Missing one-time passwords rendered as `-`** — fixed in 475f250 (+ 049aeca). Both create flows route the credentials block through a pure helper: an absent password now prints a prominent warning with the exact recovery command (`clickhousectl cloud service reset-password <uuid>` / `clickhousectl cloud postgres reset-password <uuid> --generate` — positional id, the `--id` form suggested in the finding would not parse), generic wording when the id is also absent. Exit stays 0 since the create succeeded and the password is recoverable. The follow-up fixes a contradiction: when the Postgres response omits `password` but carries `connectionString` (which the spec says embeds the password), the block points at the connection string instead of telling the user to rotate a working credential.

5. **`--json` bypassing key-material validation** — fixed in c5159bf. The `resolve_key_create_material` call is hoisted above the output-mode branch, so explicit `--json` and auto-detected agent invocations now fail before any success output; a complete response still prints the raw serialized body byte-identically (the pinned `--json` = API's actual key set policy). Wiremock tests cover incomplete-generated (both modes, non-zero exit, no success JSON), complete `--json` passthrough, and pre-hashed success.

Also in this push: 29beb7d narrows the `bind_query_endpoint` doc comment — it only ever merges the key list; the upsert still replaces `roles`/`allowedOrigins` wholesale, which is pre-existing behavior worth a separate product decision (merge vs. document).

Gates on the final head: `cargo test --workspace` (1020 passed), `cargo clippy --workspace --all-targets -- -D warnings`, `cargo fmt --all --check`, `cargo check --workspace --all-features` — all clean. A sweep for analogous spots (other `or_absent` credential renders, `unwrap_or_default` before full-replacement writes, human-branch-only validation) found none; `postgres_reset_password`'s `unwrap_or(pw)` falls back to the client-generated password it just sent, which is a real credential, not a fabrication.

**Review by iskakaushik — APPROVED** [review]
